python/ray/data/dataset_iterator.py

```diff
@@ -169,7 +178,303 @@ def iter_torch_batches(
         Returns:
             An iterator over Torch Tensor batches.
         """
-        raise NotImplementedError
+
+        # Automatically move torch tensors to the appropriate device.
+        if device is None:
```

Comment:
@amogkam , I also moved this code here. Is there any downside (dependency issue or other thing like that)? We could also inject it specifically for Train, but it seemed cleaner to just unconditionally try to get the train device.

Reply:
What was wrong with TrainIterator? It would also help with backwards compatibility

Reply:
I think we also wanted this abstraction for other Ray Train specific optimizations like overlapping host->device transfer with training.

Not sure these optimizations make sense for DatasetIteratir in general, when not used with training

Reply:
Maintaining the indirection is expensive. Also, the backwards compatibility can be handled directly in this class.

Reply:
Hmm what about the indirection is expensive? It ultimately ends up forwarding all non-implemented methods to the underlying iterator.

Reply:
You have to maintain that separate class, and keep the signatures updated, docstrings, etc. The naive forwarding isn't great.

Reply:
Any hardware acceleration features should be native to Datasets and not train, to make it general enough for other workloads too.

Reply:
I disagree- it doesn't seem like good abstractions to leak training specific optimizations into a general purpose iterator. This is why we made this abstraction in the first place https://github.com/ray-project/ray/pull/31745#discussion_r1080554006 cc @matthewdeng 


> keep the signatures updated, docstrings

I mean we already have to do this with `dataset.iter_torch_batches()` and `dataset_iterator.iter_torch_batches()` right?

> Any hardware acceleration features should be native to Datasets and not train

sure, the capabilities can be added to base DatasetIterator. But whether we turn these on by default or not may depend on whether it is being used with Ray Train or not.

Reply:
For example, with this change, any changes to `train.torch.get_device()` will now leak into Datasets-only users.

Reply:
Actually, the signature and docstrings are not an issue. It still implements `DatasetIterator` interface, and the signature and docstring are defined in the interface.

Reply:
That's true, but you still have to create a wrapper for each method in DatasetIterator, and keep these classes in sync.

How about we move `get_device` to a common location instead then? It seems to be using pure Torch+Ray logic, nothing that's actually specific to Train.

We should be detecting the right torch device for users not using Train as well.

Reply:
> That's true, but you still have to create a wrapper for each method in DatasetIterator

no you don't. It's all handled through `__getattr__`.

> How about we move get_device to a common location instead then? It seems to be using pure Torch+Ray logic, nothing that's actually specific to Train.

no, the logic is specific to how we handle `CUDA_VISIBLE_DEVICES` in train

Reply:
If there's a compelling reason to make this change, then I'm open to it. But if you're concerned about maintenance, I'll take responsibility for maintaining this.

Reply:
I believe the new implementation is very clean.

This seems generic, no?

```
def get_device() -> torch.device:
    """Gets the correct torch device to use for training.

    Assumes that `CUDA_VISIBLE_DEVICES` is set and is a
    superset of the `ray.get_gpu_ids()`.
    """
    if torch.cuda.is_available():
        # GPU IDs are assigned by Ray after you specify "use_gpu"
        # GPU `ray.get_gpu_ids()` may return ints or may return strings.
        # We should always convert to strings.
        gpu_ids = [str(id) for id in ray.get_gpu_ids()]

        if len(gpu_ids) > 0:
            # By default, there should only be one GPU ID if `use_gpu=True`.
            # If there are multiple GPUs, use the first one.
            # If using fractional GPUs, these IDs are not guaranteed
            # to be unique across different processes.
            gpu_id = gpu_ids[0]

            cuda_visible_str = os.environ.get("CUDA_VISIBLE_DEVICES", "")
            if cuda_visible_str and cuda_visible_str != "NoDevFiles":
                cuda_visible_list = cuda_visible_str.split(",")
                device_id = cuda_visible_list.index(gpu_id)
            else:
                raise RuntimeError(
                    "CUDA_VISIBLE_DEVICES set incorrectly. "
                    f"Got {cuda_visible_str}, expected to include {gpu_id}. "
                    "Did you override the `CUDA_VISIBLE_DEVICES` environment"
                    " variable? If not, please help file an issue on Github."
                )
        else:
            # If called on the driver or outside of Ray Train, return the
            # 0th device.
            device_id = 0
        device = torch.device(f"cuda:{device_id}")
    else:
        device = torch.device("cpu")

    return device
```

Reply:
ah right this is pretty generic...ok sound fine. we can add back later if needed for other Ray Train specific optimizations.

but lets still keep the backwards compatibility in base `DatasetIterator`?

Reply:
It should be backwards compatible--- I only removed an internal method.

Reply:
sorry I mean this logic: 

```
def __getattr__(self, name):
        if name == "_dataset_iterator":
            raise AttributeError

        if hasattr(self._dataset_iterator, name):
            return getattr(self._dataset_iterator, name)

        # Warning for backwards compatibility.
        warnings.warn(
            "session.get_dataset_shard returns a ray.data.DatasetIterator "
            "instead of a Dataset/DatasetPipeline as of Ray v2.3. "
            "Use iter_torch_batches(), to_tf(), or iter_batches() to "
            "iterate over one epoch. See "
            "https://docs.ray.io/en/latest/data/api/dataset_iterator.html "
            "for full DatasetIterator docs."
        )

        return getattr(self._base_dataset_or_pipeline, name)
```

Reply:
we can probably upgrade to raising an error instead of a warning for 2.4

Reply:
Oh, that warning is present for `iter_epochs()` in dataset_iterator already. I think this is probably good enough, given this is a beta API.

Reply:
but dude even a message like this `"session.get_dataset_shard returns a ray.data.DatasetIterator "` is specific to Ray Train usage.

I feel like this abstraction makes life easier for future Ray Train development...I'm ok removing if necessary, but still would prefer to keep

Reply:
Let's end this discussion here and revisit in the future.

Reply:
Hey man, I actually have some concerns about doing this automatically in Dataset iterator, specifically, I think we should honor user's choice and leave the tensor on cpu if device is not set.
For 1, this matches PyTorch's mental model and default behavior if you don't explicitly call ``to(device)``.
For 2, you can't assume this is the right time to load things onto GPU, especially now that we support ``collate_fn()``. It's quite common there are still final processing and zipping before batches are trained on (if you look at huggingface's dream booth examples).
For 3, I think scripts should run deterministically. Doing this makes me feel like users will run into confusing bugs like this runs fine on my laptop, but breaks as soon as it gets submitted to a gpu cloud instance.
I don't have context or opinion about code structure / where this should live. but I would strongly suggest we don't automatically fall back to non cpu devices here.

Reply:
In that case, is the proposal to deprecate this behavior? I don't think iterator should behave differently in Train and in other settings.

Reply:
yeah. if my understanding is correct, these 2 behave differently today, right?
I propose we carry DatasetIterator's behavior into Train, instead of the other way around, which is the change here.
wdyt?

Reply:
the reason to do it automatically is because of situations like this: https://github.com/ray-project/ray/issues/31684.

for most users, it's not obvious they need to specify it in `iter_torch_batches`, so automatically doing it for them leads to better UX.

For advanced users who are using `collate_fn`, they have to handle device movement themselves and we do not do it for them: https://github.com/ray-project/ray/blob/master/python/ray/data/dataset.py#L3004-L3009

Reply:
+1 to Amog. Would it be reasonable to keep the current approach, but make it more explicit in the API by setting the default value from `device=None` (which can be misinterpreted as no device movement) to `device="auto"`?